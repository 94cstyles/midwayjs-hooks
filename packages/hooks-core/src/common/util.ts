import parseArgs from 'fn-args'

export function isHooksMiddleware(fn: (...args: any[]) => any) {
  return parseArgs(fn).length === 1
}

export function extractMetadata(target: any) {
  const metadata: any = {}
  const metaKeys = Reflect.getMetadataKeys(target)
  for (const key of metaKeys) {
    metadata[key] = Reflect.getMetadata(key, target)
  }
  return metadata
}
