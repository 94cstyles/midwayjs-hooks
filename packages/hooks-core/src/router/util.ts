import { ServerRouter } from './'
import { extname, relative } from 'upath'
import _ from 'lodash'

export function getFunctionId(config: {
  router: ServerRouter
  sourceFilePath: string
  functionName: string
  isExportDefault: boolean
}) {
  const { router, sourceFilePath, functionName, isExportDefault } = config

  const rule = router.getRouteConfigBySourceFilePath(sourceFilePath)
  const lambdaDirectory = router.getApiDirectory(rule.baseDir)

  const length = router.config.routes.length
  // 多个 source 的情况下，根据各自的 lambdaDirectory 来增加前缀命名
  const relativeDirectory = length > 1 ? router.source : lambdaDirectory
  const relativePath = relative(relativeDirectory, sourceFilePath)
  // a/b/c -> a-b-c
  const id = _.kebabCase(removeExtension(relativePath))
  const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
  return name.toLowerCase()
}

export function removeExtension(file: string) {
  return file.replace(extname(file), '')
}
