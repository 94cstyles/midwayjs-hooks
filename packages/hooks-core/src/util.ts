import parseArgs from 'fn-args'
import last from 'lodash/last'
import { relative, resolve, sep } from 'path'

export function isDevelopment() {
  if (
    process.env.MIDWAY_TS_MODE === 'true' ||
    /* istanbul ignore next */
    process.env.NODE_ENV === 'test' ||
    /* istanbul ignore next */
    process.env.NODE_ENV === 'development'
  ) {
    return true
  }

  /* istanbul ignore next */
  return false
}

export function useHooksMiddleware(fn: (...args: any[]) => any) {
  return (...args: any[]) => {
    /**
     * @description Hooks middleware
     * @example const middleware = (next) => { const ctx = useContext() }
     */
    if (isHooksMiddleware(fn)) {
      const next = last(args)
      return fn(next)
    }
    return fn(...args)
  }
}

export function isHooksMiddleware(fn: (...args: any[]) => any) {
  return parseArgs(fn).length === 1
}

export function isPathInside(child: string, parent: string) {
  const relation = relative(parent, child)
  return Boolean(
    relation &&
      relation !== '..' &&
      !relation.startsWith(`..${sep}`) &&
      relation !== resolve(child)
  )
}

export function extractMetadata(target: any) {
  const metadata: any = {}
  const metaKeys = Reflect.getMetadataKeys(target)
  for (const key of metaKeys) {
    metadata[key] = Reflect.getMetadata(key, target)
  }
  return metadata
}
