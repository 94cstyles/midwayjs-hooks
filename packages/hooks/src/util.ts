import createDebug from 'debug'

export const debug = createDebug('hooks')
