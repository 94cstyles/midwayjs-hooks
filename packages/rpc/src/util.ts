import type { RequestArgs, RequestRoute } from '@midwayjs/hooks-core'

export function parseRequestArgs<T, I = void>(requestArgs: RequestArgs<T, I>) {
  const route = requestArgs[requestArgs.length - 1] as RequestRoute<T>
  if (route.useInputMetadata) {
    return {
      route,
      inputMetadata: requestArgs[requestArgs.length - 2] as I,
      args: requestArgs.slice(0, requestArgs.length - 2),
    }
  }

  return {
    route,
    inputMetadata: null,
    args: requestArgs.slice(0, requestArgs.length - 1),
  }
}

export function buildRPCArgs(args: any[]) {
  if (!Array.isArray(args)) {
    throw new Error('args must be an array')
  }

  return {
    args,
  }
}
