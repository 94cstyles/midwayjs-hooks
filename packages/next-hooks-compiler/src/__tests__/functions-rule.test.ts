import path from 'upath'
import { compileHooks } from './util'
import { getFunctionsMeta } from '../routes'
import { wrap } from 'jest-snapshot-serializer-raw'

describe('functions rule support', () => {
  beforeAll(async () => {
    const fixture = path.resolve(__dirname, './fixtures/functions-rule')
    await compileHooks(fixture)
  })

  it('routes', () => {
    expect(wrap(JSON.stringify(getFunctionsMeta(), null, 2))).toMatchSnapshot()
  })
})
