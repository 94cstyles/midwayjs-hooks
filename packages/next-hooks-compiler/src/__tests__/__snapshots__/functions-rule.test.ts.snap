// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler FunctionsRule Routes should be generated according to the configured rules 1`] = `
Object {
  "lambda-index": Object {
    "deployName": "lambda-index",
    "event": Object {
      "http": Object {
        "basePath": "/api",
      },
    },
    "exportFunction": "",
    "gatewayConfig": Object {
      "meta": Object {
        "functionName": "lambda-index",
      },
      "method": "GET",
      "url": "/api",
    },
    "handler": "lambda-index.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/index.js",
  },
  "render-index": Object {
    "deployName": "render-index",
    "event": Object {
      "http": Object {
        "basePath": "/",
      },
    },
    "exportFunction": "",
    "gatewayConfig": Object {
      "meta": Object {
        "functionName": "render-index",
      },
      "method": "GET",
      "url": "/",
    },
    "handler": "render-index.handler",
    "isFunctional": true,
    "sourceFilePath": "render/index.js",
  },
}
`;
